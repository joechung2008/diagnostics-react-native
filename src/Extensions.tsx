import React, { useMemo } from 'react';
import { ScrollView, Text, TouchableOpacity } from 'react-native';
import { byKey, isExtensionInfo, toNavLink } from '@/lib/utils';
import { useTheme } from './ThemeContext';

const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const { colors } = useTheme();

  const links = useMemo(
    () =>
      Object.values(extensions)
        .filter(isExtensionInfo)
        .map(toNavLink)
        .sort(byKey),
    [extensions],
  );

  const dynamicStyles = {
    container: {
      flexGrow: 0,
      flexShrink: 1,
      minWidth: 0,
      padding: 10,
      backgroundColor: colors.background,
    },
    navButton: {
      alignItems: 'center' as const,
      borderRadius: 4,
      justifyContent: 'center' as const,
      padding: 16,
      backgroundColor: colors.surface,
      borderWidth: 1,
      borderColor: colors.border,
      marginBottom: 8,
    },
    buttonText: {
      fontSize: 16,
      color: colors.text,
      textAlign: 'center' as const,
    },
  };

  return (
    <ScrollView style={dynamicStyles.container} accessibilityLabel="Extensions">
      {links.map(link => (
        <TouchableOpacity
          key={link.key}
          style={dynamicStyles.navButton}
          onPress={() => onLinkClick?.(link)}
        >
          <Text style={dynamicStyles.buttonText}>{link.name}</Text>
        </TouchableOpacity>
      ))}
    </ScrollView>
  );
};

export default Extensions;
