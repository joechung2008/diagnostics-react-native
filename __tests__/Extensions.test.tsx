import Extensions from '@/Extensions';
import { render, screen } from '@testing-library/react-native';
import React from 'react';

// Mock the ThemeContext
jest.mock('@/ThemeContext', () => ({
  useTheme: () => ({
    colors: {
      background: '#ffffff',
      surface: '#f8f9fa',
      text: '#212529',
      border: '#dee2e6',
    },
  }),
}));

describe('Extensions', () => {
  const mockOnLinkClick = jest.fn();

  it('renders correctly with multiple extensions', () => {
    const extensions = {
      'extension-a': {
        extensionName: 'extension-a',
        manageSdpEnabled: true,
        config: { enabled: 'true' },
      },
      'extension-b': {
        extensionName: 'extension-b',
        manageSdpEnabled: false,
      },
      'extension-c': {
        extensionName: 'extension-c',
        manageSdpEnabled: true,
        stageDefinition: {
          production: ['prod1'],
        },
      },
    };

    render(
      <Extensions extensions={extensions} onLinkClick={mockOnLinkClick} />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with single extension', () => {
    const extensions = {
      'single-extension': {
        extensionName: 'single-extension',
        manageSdpEnabled: true,
      },
    };

    render(
      <Extensions extensions={extensions} onLinkClick={mockOnLinkClick} />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with extension errors (filtered out)', () => {
    const extensions = {
      'good-extension': {
        extensionName: 'good-extension',
        manageSdpEnabled: true,
      },
      'bad-extension': {
        lastError: {
          errorMessage: 'Something went wrong',
          time: '2023-01-01T00:00:00Z',
        },
      },
    };

    render(
      <Extensions extensions={extensions} onLinkClick={mockOnLinkClick} />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with empty extensions', () => {
    const extensions = {};

    render(
      <Extensions extensions={extensions} onLinkClick={mockOnLinkClick} />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly without onLinkClick handler', () => {
    const extensions = {
      'test-extension': {
        extensionName: 'test-extension',
        manageSdpEnabled: true,
      },
    };

    render(
      <Extensions extensions={extensions} onLinkClick={mockOnLinkClick} />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
