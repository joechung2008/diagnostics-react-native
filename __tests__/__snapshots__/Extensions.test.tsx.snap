// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Extensions renders correctly with empty extensions 1`] = `
<RCTScrollView
  accessibilityLabel="Extensions"
  style={
    {
      "backgroundColor": "#ffffff",
      "flexGrow": 0,
      "flexShrink": 1,
      "minWidth": 0,
      "padding": 10,
    }
  }
>
  <View />
</RCTScrollView>
`;

exports[`Extensions renders correctly with extension errors (filtered out) 1`] = `
<RCTScrollView
  accessibilityLabel="Extensions"
  style={
    {
      "backgroundColor": "#ffffff",
      "flexGrow": 0,
      "flexShrink": 1,
      "minWidth": 0,
      "padding": 10,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        good-extension
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Extensions renders correctly with multiple extensions 1`] = `
<RCTScrollView
  accessibilityLabel="Extensions"
  style={
    {
      "backgroundColor": "#ffffff",
      "flexGrow": 0,
      "flexShrink": 1,
      "minWidth": 0,
      "padding": 10,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        extension-a
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        extension-b
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        extension-c
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Extensions renders correctly with single extension 1`] = `
<RCTScrollView
  accessibilityLabel="Extensions"
  style={
    {
      "backgroundColor": "#ffffff",
      "flexGrow": 0,
      "flexShrink": 1,
      "minWidth": 0,
      "padding": 10,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        single-extension
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Extensions renders correctly without onLinkClick handler 1`] = `
<RCTScrollView
  accessibilityLabel="Extensions"
  style={
    {
      "backgroundColor": "#ffffff",
      "flexGrow": 0,
      "flexShrink": 1,
      "minWidth": 0,
      "padding": 10,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f8f9fa",
          "borderColor": "#dee2e6",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 8,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        test-extension
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
